modyn/model_storage/internal/model_storage_manager.py

```diff
@@ -0,0 +1,280 @@
+        ):
+            parent_model_id: Optional[int] = self._get_parent_model_id(pipeline_id, trigger_id)
+            if parent_model_id is not None:
+                # store the model according to the incremental model strategy.
+                parent_model_state = self._get_base_model_state(pipeline_id)
+
+                # load model state of the parent model.
+                parent_model_state = self._reconstruct_model_state(parent_model_id, parent_model_state, policy)
+
+                # finally store the model delta.
+                policy.incremental_model_strategy.store_model(state_dict, parent_model_state, model_path)
+
+                return parent_model_id
+            logger.warning("Previous model is not available! Storing full model...")
+
+        # store the model in its entirety.
+        policy.full_model_strategy.store_model(state_dict, model_path)
+        return None
+
+    def _get_base_model_state(self, pipeline_id: int) -> dict:
+        """
+        Get a randomly initialized model associated with the pipeline.
+
+        Args:
+            pipeline_id: the involved pipeline.
+
+        Returns:
+            dict: the plain model state derived from the model architecture of the pipeline's models.
+        """
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            model_class_name, model_config, amp = database.get_model_configuration(pipeline_id)
+        model_module = dynamic_module_import("modyn.models")
+        assert hasattr(model_module, model_class_name), f"Model {model_class_name} not available."
+
+        model_handler = getattr(model_module, model_class_name)
+        return model_handler(json.loads(model_config), "cpu", amp).model.state_dict()
+
+    def _reconstruct_model_state(self, model_id: int, model_state: dict, policy: ModelStoragePolicy) -> dict:
+        """
+        Reconstruct a given model according to the model storage policy.
+        The function recursively calls itself whenever the model is stored as a delta.
+        Otherwise it is stored according to a full model strategy and the model state can be retrieved.
+        Finally, the model_state is overwritten by the state of the inquired model.
+
+        Args:
+            model_id: the identifier of the model to be reconstructed.
+            model_state: a random model state (or the loaded parent model state).
+            policy: the model storage policy of the pipeline.
+        Returns:
+            dict: the reconstructed model state. Refers to the same object as model_state.
+        """
+
+        # we recursively overwrite the model state.
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            model: TrainedModel = database.session.get(TrainedModel, model_id)
+        if not model.parent_model:
+            # base case: we can load a fully stored model.
+            policy.full_model_strategy.load_model(model_state, self._storage_dir / model.model_path)
```

Comment:
Can we also change to use return values here? (i.e., the function also returns `model_state` and we assign it). Just to make the update more explicit